### ISSUE TITLE
Question about the mixup operation in this code

### ISSUE THREAD

**Issue by bad-meets-joke** [mancinimassimiliano/CuMix#2]
Hello, I think there exsists a logical error in your code. 

In `methods.py`,   the class method `get_sample_mixup()` of the class `CuMix` is **not consistant with the mixup rules your proposed in paper**.

Here I take out the error line 246

> indeces = ratio_within_dom * permuted_within_dom + (1. - ratio_within_dom) * permuted_across_dom

In your code, you mixup the index instead of content (image or feature) or label,  then the mixup index are employed to extract content from current batch.

This is not reasonable and not consistant with `Eq. (5)` in your paper.

Hope you can check the code. Thanks

**Comment by mancinimassimiliano** [comment]
Hi @bad-meets-joke,

As Eq. (5) shows, our mixup is implemented in 2 steps. First, we decide whether to perform mixup within the same or different domain and then we perform the actual mixing of features and images. For the latter, we do it  [here](https://github.com/mancinimassimiliano/CuMix/blob/8a1cf3a5071e12ca4eee0fcf2e55edb9fdef48f4/methods.py#L260), after sampling the mixing ratio from a beta distribution [here](https://github.com/mancinimassimiliano/CuMix/blob/8a1cf3a5071e12ca4eee0fcf2e55edb9fdef48f4/methods.py#L252).  

For deciding which samples to mix instead (i.e., within- or cross-domain), we say in the paper that we use a Bernoulli distribution (just below Eq. (5) ). Sampling from a Bernoulli distribution returns either 0 or 1, i.e., we do not use a ratio in the domain part, but we pick one of the two samples (you can see the returned values as an on/off signal). This is why we state: _Note that we
introduced a term γ which allows performing either intra-domain (with γ = 0)
or cross-domain (with γ = 1) mixes._

Having clarified this, mixing the indices is just a shortcut that I used to write a (hopefully) cleaner code, while changing **nothing** from an outcome perspective. You can easily check this with this snippet of code:

```
import torch
import numpy as np
RG = np.random.default_rng()

def sanity_check(a,b,sampling_beta,sampling_gamma):
	# Sample beta ratio and domain indicator
	beta = torch.from_numpy(RG.beta(sampling_beta,sampling_beta,a.size(0))).unsqueeze(1)
	gamma = torch.from_numpy(RG.binomial(1,sampling_gamma,a.size(0)))

	# Randomly permute the indices for the second tensor, as a placeholder for cross-domain mixes
	cross_domain = torch.randperm(a.size(0))

	# Just use ordered indices, placeholder for same domain mixing
	same_domain = torch.arange(a.size(0))

	# Mix as Eq.(5)
	mix_eq = beta*a+(1-beta)*(gamma.unsqueeze(-1)*b[same_domain] + (1-gamma.unsqueeze(-1))*b[cross_domain])

	# Mix as in code
	domain_idxs = gamma*same_domain + (1-gamma)*cross_domain
	mix_code = beta*a+(1-beta)*b[domain_idxs]

	# Check difference
	diff = torch.abs(mix_code-mix_eq).sum()
	return diff==0

BS=100
FEATURES=1000
SAMPLING_BETA = 1.0
SAMPLING_GAMMA = 0.5

a = torch.FloatTensor(BS,FEATURES).uniform_()
b = torch.FloatTensor(BS,FEATURES).uniform_()

print(sanity_check(a,b,SAMPLING_BETA,SAMPLING_GAMMA))

```
I am sorry if the paper and/or the implementation results somewhat unclear on the way we perform cross-domain mixes. I hope with this answer I clarified that the two are exactly the same.

**Comment by bad-meets-joke** [comment]
Thanks for your timely answer@mancinimassimiliano. 

**You are right. There is no error in code.**  

I do neglect the fact that sampling results γ from Bernoulli distribution are binary.
